sdk/src/error.rs

```diff
@@ -1,33 +1,78 @@
-use anchor_lang::prelude::error_code;
+use light_hasher::HasherError;
+use solana_program::program_error::ProgramError;
+use thiserror::Error;
 
-#[error_code]
+#[derive(Debug, Error)]
 pub enum LightSdkError {
-    #[msg("Constraint violation")]
+    #[error("Constraint violation")]
     ConstraintViolation,
-    #[msg("Invalid light-system-program ID")]
+    #[error("Invalid light-system-program ID")]
     InvalidLightSystemProgram,
-    #[msg("Expected accounts in the instruction")]
+    #[error("Expected accounts in the instruction")]
     ExpectedAccounts,
-    #[msg("Expected address Merkle context to be provided")]
+    #[error("Expected address Merkle context to be provided")]
     ExpectedAddressMerkleContext,
-    #[msg("Expected address root index to be provided")]
+    #[error("Expected address root index to be provided")]
     ExpectedAddressRootIndex,
-    #[msg("Accounts with a specified input are expected to have data")]
+    #[error("Accounts with a specified input are expected to have data")]
     ExpectedData,
-    #[msg("Accounts with specified data are expected to have a discriminator")]
+    #[error("Accounts with specified data are expected to have a discriminator")]
     ExpectedDiscriminator,
-    #[msg("Accounts with specified data are expected to have a hash")]
+    #[error("Accounts with specified data are expected to have a hash")]
     ExpectedHash,
-    #[msg("`mut` and `close` accounts are expected to have a Merkle context")]
+    #[error("`mut` and `close` accounts are expected to have a Merkle context")]
     ExpectedMerkleContext,
-    #[msg("Expected root index to be provided")]
+    #[error("Expected root index to be provided")]
     ExpectedRootIndex,
-    #[msg("Cannot transfer lamports from an account without input")]
+    #[error("Cannot transfer lamports from an account without input")]
     TransferFromNoInput,
-    #[msg("Cannot transfer from an account without lamports")]
+    #[error("Cannot transfer from an account without lamports")]
     TransferFromNoLamports,
-    #[msg("Account, from which a transfer was attempted, has insufficient amount of lamports")]
+    #[error("Account, from which a transfer was attempted, has insufficient amount of lamports")]
     TransferFromInsufficientLamports,
-    #[msg("Integer overflow resulting from too large resulting amount")]
+    #[error("Integer overflow resulting from too large resulting amount")]
     TransferIntegerOverflow,
+
+    #[error(transparent)]
+    Borsh(#[from] borsh::maybestd::io::Error),
+    #[error(transparent)]
+    Hasher(#[from] HasherError),
+}
+
+impl From<LightSdkError> for u32 {
+    fn from(e: LightSdkError) -> Self {
+        match e {
+            LightSdkError::ConstraintViolation => 13001,
```

Comment:
we already use 13000 + for errors of the `light-verifier` crate, can we use 14000 instead?